### PR TITLE
fix Gemma4 prepare() to honor windowSize -- prompt prefill is never chunked

### PR DESCRIPTION
**Pull request by joshuaejackson**
## Proposed changes
Addresses the Gemma4 instance of #336 (MLXVLM models ignore `windowSize` — same fix shape as #297).

`Gemma4.prepare(_:cache:windowSize:)` accepts `windowSize` but never uses it: the whole prompt runs through the language model in one forward pass, so transient prefill memory scales with prompt length (~1.7 MB/token measured on E4B 4-bit — a 4k-token prompt peaks at 11.2 GB MLX active against 4.3 GiB of weights; 16k crashes an 18 GB machine).

This applies the same chunking #297 added for Idefics3, to both Gemma4 branches: the text path chunks raw tokens; the image path chunks the merged embeddings together with the aligned per-layer-input slices. The KV cache is evaluated between chunks; the final position returns the first-token logits.

Gemma4-specific notes: KV-shared layers read the source layer's cache-augmented keys/values, so per-chunk forwards still see full history; per-layer inputs are per-token; RoPE offsets and masks derive from `cache.offset`. 

## Verification (Gemma4 E4B 4-bit, M3 Pro 18 GB)

- **Memory** (5 iters/point, temp 0, medians): peak MLX active 11.2→5.0 GB at 4k, 17.7→5.1 GB at 8k; 16k was un-runnable before (~33 GB projected) and peaks at 5.2 GB after. TTFT improved 25.6→17.9 s (4k) and 70.5→47.1 s (8k).
- **Output equality:** greedy outputs byte-identical before/after at 34/492/573/733-token prompts (spanning the 512 chunk boundary). At 4k/8k, long free-form greedy generations are deterministic per build and can diverge at near-tie tokens (floating-point evaluation order — same class of variation as changing `prefillStepSize` on `LLMModel`); needle-recall scoring was 5/5 in both configurations at all lengths.

## Checklist

Put an `x` in the boxes that apply.

- [x] I have read the [CONTRIBUTING](https://github.com/ml-explore/mlx/blob/main/CONTRIBUTING.md) document
- [x] I have run `pre-commit run --all-files` to format my code / installed pre-commit prior to committing changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have updated the necessary documentation (if needed)

### CODE REVIEW

File: Libraries/MLXVLM/Models/Gemma4.swift [hunk 1]
```diff
@@ -1723,18 +1723,54 @@ public final class Gemma4: Module, VLMModel, KVCacheDimensionProvider {
...
+                let range = processed ..< (processed + chunkLength)
+                _ = languageModel(
+                    nil,
+                    cache: convertedCache,
+                    inputsEmbeds: allEmbeds[0..., range, 0...],
+                    perLayerInputs: allPerLayerInputs.map { $0[0..., range, 0..., 0...] }
+                )
+                eval(cache)
+                processed += chunkLength
+            }
```
Comment (davidkoski) [inline Libraries/MLXVLM/Models/Gemma4.swift:1748]: Since we are here touching it (and I suspect this pattern will be copied), let's use the asyncEval + eval pattern from LLMModel:

```swift
        // Prepare the prompt in chunks if larger than the prefill size.
        // asyncEval lets the CPU build chunk N+1's graph while the GPU evaluates
        // chunk N.
        var state: LMOutput.State?
        while y.tokens.size > prefillStepSize {
            let input = y[.newAxis, ..<prefillStepSize]
            let output = self(input, cache: cache.isEmpty ? nil : cache, state: state)
            state = output.state
            asyncEval(cache)
            y = y[prefillStepSize...]
        }

        // Single sync after the loop to flush any remaining async work.
        eval(cache)
```

asyncEval inside the loop and eval after

Comment (joshuaejackson) [reply Libraries/MLXVLM/Models/Gemma4.swift:1748]: Thanks!! Done. Switched to asyncEval in the loop with a single eval(cache) after, in both branches. (Gemma4 doesn't carry decoder state, so I skipped the state threading from your snippet.)

I also added a test file — it builds a tiny random-weight Gemma4 and checks that chunked and single-pass prefill give the same answer. If the pattern gets copied, the test can get copied.

**Review by davidkoski — APPROVED** [review]

**Comment by davidkoski** [comment]
Thank you for the contribution!
